crates/cashu/src/nuts/nut00.rs

```diff
@@ -366,34 +366,89 @@ pub mod wallet {
 
             (amount.into(), self.token[0].mint.to_string())
         }
+
+        pub fn to_v4_string(&self) -> String {
+            TokenV4::from(self.clone()).to_string()
+        }
     }
 
     impl FromStr for Token {
         type Err = error::wallet::Error;
 
         fn from_str(s: &str) -> Result<Self, Self::Err> {
-            let s = if s.starts_with("cashuA") {
+            let is_v3 = s.starts_with("cashuA");
+            let is_v4 = s.starts_with("cashuB");
+            let s = if is_v3 {
                 s.replace("cashuA", "")
+            } else if is_v4 {
+                s.replace("cashuB", "")
             } else {
                 return Err(wallet::Error::UnsupportedToken);
             };
 
             let decode_config = general_purpose::GeneralPurposeConfig::new()
                 .with_decode_padding_mode(base64::engine::DecodePaddingMode::Indifferent);
-            let decoded = GeneralPurpose::new(&alphabet::STANDARD, decode_config).decode(s)?;
-            let decoded_str = String::from_utf8(decoded)?;
-            let token: Token = serde_json::from_str(&decoded_str)?;
-            Ok(token)
+            let decoded = GeneralPurpose::new(&alphabet::URL_SAFE, decode_config).decode(s)?;
+            if is_v3 {
+                let decoded_str = String::from_utf8(decoded)?;
+                let token: Token = serde_json::from_str(&decoded_str)?;
+                Ok(token)
+            } else {
+                println!("{}", hex::encode(decoded.clone()));
+                let token: TokenV4 = ciborium::from_reader(&decoded[..])?;
+                Ok(token.into())
+            }
         }
     }
 
     impl fmt::Display for Token {
         fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
             let json_string = serde_json::to_string(self).map_err(|_| fmt::Error)?;
-            let encoded = general_purpose::STANDARD.encode(json_string);
+            let encoded = general_purpose::URL_SAFE.encode(json_string);
             write!(f, "cashuA{}", encoded)
         }
     }
+
+    impl From<TokenV4> for Token {
+        fn from(token: TokenV4) -> Self {
+            Token {
+                token: token.token.into_iter().map(Into::into).collect(),
+                memo: token.memo,
+                unit: token.unit,
+            }
+        }
+    }
+
+    #[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+    struct TokenV4 {
+        #[serde(rename = "t")]
+        pub token: Vec<MintProofsV4>,
+        /// Memo for token
+        #[serde(rename = "m", skip_serializing_if = "Option::is_none")]
+        pub memo: Option<String>,
+        /// Token Unit
+        #[serde(rename = "u", skip_serializing_if = "Option::is_none")]
+        pub unit: Option<CurrencyUnit>,
+    }
+
+    impl fmt::Display for TokenV4 {
+        fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            let mut data = Vec::new();
+            ciborium::into_writer(self, &mut data).unwrap();
```

Comment:
i highly discourage using `unwrap`.

Recommend going with `expect` and adding useful log to the user

Reply:
I agree unwrap should only be allowed within tests. Ideally a specific and useful error is returned